File: packages/clients/rust/src/process/stdio/read.rs
use {
	super::{Chunk, Stream},
	crate::prelude::*,
	bytes::Bytes,
	futures::{
		StreamExt as _, TryStreamExt as _,
		stream::{self, BoxStream},
	},
	num::ToPrimitive as _,
	serde_with::{DurationSecondsWithFrac, serde_as},
	std::time::Duration,
	tangram_http::{request::builder::Ext as _, response::Ext as _},
	tangram_uri::Uri,
	tangram_util::serde::{CommaSeparatedString, SeekFromNumberOrString},
};

#[serde_as]
#[derive(
	Clone,
	Debug,
	Default,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
pub struct Arg {
	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 0, skip_serializing_if = "Option::is_none")]
	pub length: Option<i64>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 1, skip_serializing_if = "Option::is_none")]
	pub location: Option<tg::location::Arg>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[serde_as(as = "Option<SeekFromNumberOrString>")]
	#[tangram_serialize(default, id = 2, skip_serializing_if = "Option::is_none")]
	pub position: Option<std::io::SeekFrom>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 3, skip_serializing_if = "Option::is_none")]
	pub size: Option<u64>,

	#[serde(default, skip_serializing_if = "Vec::is_empty")]
	#[serde_as(as = "CommaSeparatedString")]
	#[tangram_serialize(default, id = 4, skip_serializing_if = "Vec::is_empty")]
	pub streams: Vec<Stream>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[serde_as(as = "Option<DurationSecondsWithFrac>")]
	#[tangram_serialize(default, id = 5, skip_serializing_if = "Option::is_none")]
	pub timeout: Option<Duration>,

	#[serde(default, skip_serializing_if = "tg::Tokens::is_empty")]
	#[tangram_serialize(default, id = 6, skip_serializing_if = "tg::Tokens::is_empty")]
	pub tokens: tg::Tokens,
}

#[derive(
	Clone,
	Debug,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
#[serde(content = "value", rename_all = "snake_case", tag = "kind")]
pub enum ClientMessage {
	#[tangram_serialize(id = 0)]
	Ack,
	#[tangram_serialize(id = 1)]
	Notification(Progress),
}

#[derive(
	Clone,
	Debug,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
#[serde(content = "value", rename_all = "snake_case", tag = "kind")]
pub enum ServerMessage {
	#[tangram_serialize(id = 0)]
	Notification(Event),
	#[tangram_serialize(id = 1)]
	Response(Output),
}

#[derive(
	Clone,
	Debug,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
#[serde(content = "value", rename_all = "snake_case", tag = "kind")]
pub enum Event {
	#[tangram_serialize(id = 0)]
	Chunk(Chunk),
	/// The resolved cursor and remaining length, including any clipping at EOF.
	#[tangram_serialize(id = 1)]
	Position {
		#[tangram_serialize(id = 0)]
		length: Option<i64>,
		#[tangram_serialize(id = 1)]
		position: u64,
	},
}

#[derive(Clone, Debug, Default)]
pub struct Options {
	pub length: Option<i64>,
	pub location: Option<tg::location::Arg>,
	pub position: Option<std::io::SeekFrom>,
	pub size: Option<u64>,
	pub streams: Vec<Stream>,
	pub timeout: Option<Duration>,
}

#[derive(
	Clone,
	Debug,
	Eq,
	PartialEq,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
#[serde(content = "value", rename_all = "snake_case", tag = "kind")]
pub enum Output {
	#[tangram_serialize(id = 0)]
	End(super::End),
	#[tangram_serialize(id = 1)]
	Limit {
		#[tangram_serialize(id = 0)]
		position: u64,
	},
	#[tangram_serialize(id = 2)]
	Timeout {
		#[tangram_serialize(id = 0)]
		position: u64,
	},
}

#[derive(
	Clone,
	Copy,
	Debug,
	Default,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
pub struct Progress {
	/// The cumulative number of consumed chunk bytes in this read attempt.
	#[tangram_serialize(id = 0)]
	pub consumed: u64,
}

impl Output {
	pub fn validate(&self, streams: &[Stream], position: u64) -> tg::Result<()> {
		let expected = self.position(streams)?;
		let valid = match self {
			// A forward read may begin beyond EOF, but it must not finish short of EOF.
			Self::End(_) => position >= expected,
			Self::Limit { .. } | Self::Timeout { .. } => position == expected,
		};
		if !valid {
			return Err(
				tg::error!(expected = %expected, actual = %position, "encountered a gap at the end of the stdio read"),
			);
		}
		Ok(())
	}

	pub fn position(&self, streams: &[Stream]) -> tg::Result<u64> {
		let position = match self {
			Self::End(end) if streams.len() > 1 => end.combined_position,
			Self::End(end) => *streams
				.first()
				.and_then(|stream| end.stream_positions.get(stream))
				.ok_or_else(|| tg::error!("missing the stdio end position"))?,
			Self::Limit { position } | Self::Timeout { position } => *position,
		};
		Ok(position)
	}
}

impl<O> tg::Process<O> {
	pub async fn try_read_stdio(
		&self,
		options: tg::process::stdio::read::Options,
	) -> tg::Result<Option<BoxStream<'static, tg::Result<tg::process::stdio::Chunk>>>> {
		let handle = tg::handle()?;
		self.try_read_stdio_with_handle(handle, options).await
	}

	pub async fn try_read_stdio_with_handle<H>(
		&self,
		handle: &H,
		options: tg::process::stdio::read::Options,
	) -> tg::Result<Option<BoxStream<'static, tg::Result<tg::process::stdio::Chunk>>>>
	where
		H: tg::Handle,
	{
		if options.streams.is_empty() {
			return Err(tg::error!("expected at least one stdio stream"));
		}
		let handle = self.handle_with_handle(handle);
		let handle = &handle;
		if self.id().is_left() {
			let mut streams = Vec::<
				BoxStream<'static, tg::Result<(Bytes, tg::process::stdio::Stream, u64)>>,
			>::new();
			for stream in options.streams {
				match stream {
					tg::process::stdio::Stream::Stdin => {
						return Err(tg::error!("reading stdin is invalid"));
					},
					tg::process::stdio::Stream::Stdout => {
						let handle = handle.clone();
						let stdout = self.stdout();
						let stream = stream::try_unfold(
							(handle, stdout, 0),
							|(handle, mut stdout, stream_position)| async move {
								let Some(bytes) = stdout.read_with_handle(&handle).await? else {
									return Ok(None);
								};
								let length = bytes.len().to_u64().unwrap();
								let item =
									(bytes, tg::process::stdio::Stream::Stdout, stream_position);
								let stream_position = stream_position + length;

								Ok(Some((item, (handle, stdout, stream_position))))
							},
						);
						streams.push(stream.boxed());
					},
					tg::process::stdio::Stream::Stderr => {
						let handle = handle.clone();
						let stderr = self.stderr();
						let stream = stream::try_unfold(
							(handle, stderr, 0),
							|(handle, mut stderr, stream_position)| async move {
								let Some(bytes) = stderr.read_with_handle(&handle).await? else {
									return Ok(None);
								};
								let length = bytes.len().to_u64().unwrap();
								let item =
									(bytes, tg::process::stdio::Stream::Stderr, stream_position);
								let stream_position = stream_position + length;

								Ok(Some((item, (handle, stderr, stream_position))))
							},
						);
						streams.push(stream.boxed());
					},
				}
			}
			let stream = futures::stream::select_all(streams).boxed();
			let stream =
				stream::try_unfold((stream, 0), |(mut stream, combined_position)| async move {
					let Some((bytes, stream_name, stream_position)) = stream.try_next().await?
					else {
						return Ok(None);
					};
					let length = bytes.len().to_u64().unwrap();
					let chunk = tg::process::stdio::Chunk {
						bytes,
						combined_position,
						stream: stream_name,
						stream_position,
						timestamp: None,
					};
					let combined_position = combined_position + length;

					Ok(Some((chunk, (stream, combined_position))))
				});

			return Ok(Some(stream.boxed()));
		}

		let id = self.id().unwrap_right();
		let arg = tg::process::stdio::read::Arg {
			length: options.length,
			location: options.location.or_else(|| self.location()),
			position: options.position,
			size: options.size,
			streams: options.streams,
			timeout: options.timeout,
			tokens: self.tokens(),
		};
		let Some(stream) = handle.try_read_process_stdio_all(id, arg).await? else {
			return Ok(None);
		};

		Ok(Some(stream.boxed()))
	}
}

impl tg::Session {
	pub async fn try_read_process_stdio(
		&self,
		id: &tg::process::Id,
		arg: tg::process::stdio::read::Arg,
		input: BoxStream<'static, tg::Result<tg::process::stdio::read::ClientMessage>>,
	) -> tg::Result<
		Option<
			impl futures::Stream<Item = tg::Result<tg::process::stdio::read::ServerMessage>>
			+ Send
			+ 'static
			+ use<>,
		>,
	> {
		if arg.streams.is_empty() {
			return Err(tg::error!("expected at least one stdio stream"));
		}
		let max_frame_size = self.client().sync.max_frame_size;
		let path = format!("/processes/{id}/stdio/read");
		let uri = Uri::builder().path(&path).build().unwrap();
		let body = super::encode(input, max_frame_size);
		let request = http::request::Builder::default()
			.method(http::Method::POST)
			.uri(uri)
			.header(http::header::ACCEPT, super::TANGRAM_CONTENT_TYPE)
			.header(http::header::CONTENT_TYPE, super::TANGRAM_CONTENT_TYPE)
			.arg(&arg, body)
			.map_err(|error| tg::error!(!error, "failed to serialize the arg"))?
			.unwrap();
		let response = self
			.send(request)
			.await
			.map_err(|error| tg::error!(!error, "failed to send the request"))?;
		if response.status() == http::StatusCode::NOT_FOUND {
			return Ok(None);
		}
		if !response.status().is_success() {
			let status = response.status();
			let error = response
				.json::<tg::Error>()
				.await
				.map_err(|error| tg::error!(!error, "failed to deserialize the error response"))?;
			let error = tg::error!(!error, status = %status, "the request failed");
			return Err(error);
		}
		let content_type = response
			.parse_header::<mime::Mime, _>(http::header::CONTENT_TYPE)
			.transpose()?;
		if content_type != Some(super::TANGRAM_CONTENT_TYPE.parse().unwrap()) {
			return Err(tg::error!(?content_type, "invalid content type"));
		}
		let stream = super::decode(response.into_body(), max_frame_size);

		Ok(Some(stream))
	}
}

impl TryFrom<ClientMessage> for tangram_http::sse::Event {
	type Error = tg::Error;
	fn try_from(message: ClientMessage) -> tg::Result<Self> {
		let (event, data) = match message {
			ClientMessage::Ack => ("ack", Ok("null".to_owned())),
			ClientMessage::Notification(progress) => {
				("notification", serde_json::to_string(&progress))
			},
		};
		let data = data.map_err(|error| tg::error!(!error, "failed to serialize the message"))?;
		let event = Self {
			data,
			event: Some(event.to_owned()),
			..Default::default()
		};
		Ok(event)
	}
}

impl TryFrom<tangram_http::sse::Event> for ClientMessage {
	type Error = tg::Error;
	fn try_from(event: tangram_http::sse::Event) -> tg::Result<Self> {
		let message = match event.event.as_deref() {
			Some("ack") => serde_json::from_str::<()>(&event.data).map(|()| Self::Ack),
			Some("notification") => serde_json::from_str(&event.data).map(Self::Notification),
			Some("error") => {
				let error: tg::Either<tg::error::Data, tg::error::Id> =
					serde_json::from_str(&event.data)
						.map_err(|error| tg::error!(!error, "failed to deserialize the error"))?;
				return Err(error.try_into()?);
			},
			_ => return Err(tg::error!("invalid stdio read message")),
		}
		.map_err(|error| tg::error!(!error, "failed to deserialize the message"))?;
		Ok(message)
	}
}

impl TryFrom<ServerMessage> for tangram_http::sse::Event {
	type Error = tg::Error;
	fn try_from(message: ServerMessage) -> tg::Result<Self> {
		let (event, data) = match message {
			ServerMessage::Notification(event) => ("notification", serde_json::to_string(&event)),
			ServerMessage::Response(output) => ("response", serde_json::to_string(&output)),
		};
		let data = data.map_err(|error| tg::error!(!error, "failed to serialize the message"))?;
		let event = Self {
			data,
			event: Some(event.to_owned()),
			..Default::default()
		};
		Ok(event)
	}
}

impl TryFrom<tangram_http::sse::Event> for ServerMessage {
	type Error = tg::Error;
	fn try_from(event: tangram_http::sse::Event) -> tg::Result<Self> {
		let message = match event.event.as_deref() {
			Some("error") => {
				let error: tg::Either<tg::error::Data, tg::error::Id> =
					serde_json::from_str(&event.data)
						.map_err(|error| tg::error!(!error, "failed to deserialize the error"))?;
				return Err(error.try_into()?);
			},
			Some("notification") => serde_json::from_str(&event.data).map(Self::Notification),
			Some("response") => serde_json::from_str(&event.data).map(Self::Response),
			_ => return Err(tg::error!("invalid stdio read message")),
		}
		.map_err(|error| tg::error!(!error, "failed to deserialize the message"))?;
		Ok(message)
	}
}

File: packages/clients/js/src/client/process/stdio/read.ts
import * as tg from "../../../index.ts";
import { Body, Request, Response, Uri, percentEncode } from "../../../http.ts";
import { Receiver } from "../../../process/stdio/flow.ts";
import type { Client } from "../../../client.ts";

export type Connection = {
	input: {
		close(): void;
		push(message: tg.Process.Stdio.Read.ClientMessage): boolean;
	};
	output: AsyncIterableIterator<tg.Process.Stdio.Read.ServerMessage>;
	reconnect?: (arg: tg.Process.Stdio.Read.Arg) => Promise<Connection>;
};

class ProtocolError extends Error {}

export async function tryReadProcessStdio(
	client: Client,
	id: tg.Process.Id,
	arg: tg.Process.Stdio.Read.Arg,
): Promise<AsyncIterableIterator<tg.Process.Stdio.Chunk> | null> {
	let connection = await connect(client, id, arg);
	if (connection === null) {
		return null;
	}

	return readProcessStdioAll(client, id, arg, connection);
}

export function readProcessStdioAll(
	client: Client,
	id: tg.Process.Id,
	arg: tg.Process.Stdio.Read.Arg,
	connection: Connection,
): AsyncIterableIterator<tg.Process.Stdio.Chunk> {
	let state = { canceled: false, connection };
	let output = readProcessStdioAllInner(client, id, arg, state);
	return {
		next: () => output.next(),
		return: async () => {
			state.canceled = true;
			state.connection.input.close();
			state.connection.output.return?.().catch(() => {});
			return await output.return!();
		},
		[Symbol.asyncIterator]() {
			return this;
		},
	};
}

async function* readProcessStdioAllInner(
	client: Client,
	id: tg.Process.Id,
	arg: tg.Process.Stdio.Read.Arg,
	state: { canceled: boolean; connection: Connection },
): AsyncIterableIterator<tg.Process.Stdio.Chunk> {
	let connection = state.connection;
	let combined = arg.streams.length > 1;
	let forward =
		arg.length === undefined || arg.length === null || arg.length >= 0;
	let nextArg = { ...arg, streams: [...arg.streams] };
	let window = new Receiver();
	let pending = 0;
	let position = typeof arg.position === "string" ? null : (arg.position ?? 0);
	try {
		while (!state.canceled) {
			let progress = window.consume(pending);
			pending = 0;
			if (progress !== null) {
				connection.input.push({ kind: "notification", value: progress });
			}
			let result: IteratorResult<tg.Process.Stdio.Read.ServerMessage> | null =
				null;
			try {
				result = await connection.output.next();
			} catch (error) {
				if (state.canceled) return;
				if (isTerminalError(error)) {
					throw error;
				}
			}
			if (state.canceled) return;
			if (result === null || result.done) {
				connection = state.connection = await reconnect(
					client,
					id,
					nextArg,
					connection,
				);
				window = new Receiver();
				continue;
			}
			let message = result.value;
			if (message.kind === "response") {
				tg.Process.Stdio.Read.Output.validate(
					message.value,
					arg.streams,
					position ?? 0,
				);
				connection.input.push({ kind: "ack" });
				connection.input.close();
				return;
			}
			if (message.value.kind === "position") {
				let value = message.value.value;
				if (
					!Number.isSafeInteger(value.position) ||
					value.position < 0 ||
					(value.length !== null && !Number.isSafeInteger(value.length))
				) {
					throw new ProtocolError("invalid process stdio position");
				}
				position = value.position;
				nextArg.position = position;
				nextArg.length = value.length;
				continue;
			}
			if (message.value.kind !== "chunk") {
				throw new ProtocolError("invalid process stdio read notification");
			}
			let chunk = message.value.value;
			pending = chunk.bytes.length;
			if (!arg.streams.includes(chunk.stream)) {
				throw new ProtocolError("invalid process stdio stream");
			}
			let start = combined ? chunk.combinedPosition : chunk.streamPosition;
			let end = start + chunk.bytes.length;
			if (!Number.isSafeInteger(end)) {
				throw new ProtocolError("the stdio position is too large");
			}
			if (position !== null) {
				if ((forward && end <= position) || (!forward && start >= position)) {
					continue;
				}
				if ((forward && start > position) || (!forward && end < position)) {
					throw new ProtocolError("encountered a gap in the stdio stream");
				}
				if (forward && start < position) {
					let overlap = position - start;
					chunk = {
						...chunk,
						bytes: chunk.bytes.subarray(overlap),
						combinedPosition: chunk.combinedPosition + overlap,
						streamPosition: chunk.streamPosition + overlap,
					};
				} else if (!forward && end > position) {
					chunk = {
						...chunk,
						bytes: chunk.bytes.subarray(0, position - start),
					};
				}
			}
			let length = chunk.bytes.length;
			position = forward
				? (combined ? chunk.combinedPosition : chunk.streamPosition) + length
				: combined
					? chunk.combinedPosition
					: chunk.streamPosition;
			if (nextArg.length !== undefined && nextArg.length !== null) {
				if (nextArg.length >= 0) {
					nextArg.length -= Math.min(length, nextArg.length);
				} else {
					nextArg.length += Math.min(length, Math.abs(nextArg.length));
				}
			}
			nextArg.position = position;

			yield chunk;
		}
	} finally {
		connection.input.close();
		await connection.output.return?.();
	}
}

async function connect(
	client: Client,
	id: tg.Process.Id,
	arg: tg.Process.Stdio.Read.Arg,
): Promise<Connection | null> {
	let attempt = 0;
	while (true) {
		try {
			return await readProcessStdioOnce(client, id, arg);
		} catch (error) {
			if (isTerminalError(error)) {
				throw error;
			}
			await retryDelay(attempt);
			attempt++;
		}
	}
}

async function reconnect(
	client: Client,
	id: tg.Process.Id,
	arg: tg.Process.Stdio.Read.Arg,
	connection: Connection,
): Promise<Connection> {
	connection.input.close();
	await connection.output.return?.();
	let next =
		connection.reconnect === undefined
			? await connect(client, id, arg)
			: await connection.reconnect(arg);
	if (next === null) {
		throw new Error("failed to find the process");
	}

	return next;
}

async function readProcessStdioOnce(
	client: Client,
	id: tg.Process.Id,
	arg: tg.Process.Stdio.Read.Arg,
): Promise<Connection | null> {
	let input = new Channel<tg.Process.Stdio.Read.ClientMessage>();
	let uri = new Uri({
		path: `/processes/${percentEncode(id)}/stdio/read`,
		query: {
			...arg,
			location:
				arg.location === undefined || arg.location === null
					? null
					: tg.Location.Arg.toDataString(arg.location),
			streams: arg.streams.join(","),
		},
	});
	let request = new Request({
		body: Body.sse(encodeClientMessages(input)),
		headers: {
			accept: "text/event-stream",
			"content-type": "text/event-stream",
		},
		method: "POST",
		uri,
	});
	let response = await client.send(request);
	if (response.status === 404) {
		input.close();
		return null;
	}
	if (response.status < 200 || response.status >= 300) {
		input.close();
		throw await responseError(response);
	}
	let contentType = response.headers.get("content-type")?.split(";", 1)[0];
	if (contentType !== "text/event-stream") {
		input.close();
		throw new ProtocolError("invalid process stdio response content type");
	}
	let output = decodeServerMessages(response);

	return { input, output };
}

async function* encodeClientMessages(
	input: AsyncIterable<tg.Process.Stdio.Read.ClientMessage>,
): AsyncIterableIterator<Body.SseEvent> {
	for await (let message of input) {
		yield {
			data: JSON.stringify(message.kind === "ack" ? null : message.value),
			event: message.kind,
		};
	}
}

async function* decodeServerMessages(
	response: Response,
): AsyncIterableIterator<tg.Process.Stdio.Read.ServerMessage> {
	for await (let event of response.sse()) {
		try {
			if (event.event === "error") {
				throw errorFromData(
					JSON.parse(event.data) as tg.Error.Data | tg.Error.Id,
				);
			}
			if (event.event === "notification") {
				let value = JSON.parse(event.data) as
					| { kind: "chunk"; value: tg.Process.Stdio.Chunk.Data }
					| {
							kind: "position";
							value: { length: number | null; position: number };
					  };
				if (value.kind === "chunk") {
					yield {
						kind: "notification",
						value: {
							kind: "chunk",
							value: tg.Process.Stdio.Chunk.fromData(value.value),
						},
					};
				} else if (value.kind === "position") {
					yield { kind: "notification", value };
				} else {
					throw new ProtocolError("invalid process stdio read notification");
				}
			} else if (event.event === "response") {
				let value = tg.Process.Stdio.Read.Output.fromData(
					JSON.parse(event.data),
				);
				if (!["end", "limit", "timeout"].includes(value.kind)) {
					throw new ProtocolError("invalid process stdio read request");
				}
				yield { kind: "response", value };
			} else {
				throw new ProtocolError("invalid process stdio read message");
			}
		} catch (error) {
			if (error instanceof tg.Error || error instanceof ProtocolError) {
				throw error;
			}
			throw new ProtocolError("failed to deserialize a process stdio message", {
				cause: error,
			});
		}
	}
}

async function responseError(response: Response): Promise<tg.Error> {
	try {
		return errorFromData(await response.json<tg.Error.Data | tg.Error.Id>());
	} catch (error) {
		if (error instanceof tg.Error) {
			throw error;
		}
		throw new ProtocolError("failed to deserialize the error response", {
			cause: error,
		});
	}
}

function errorFromData(data: tg.Error.Data | tg.Error.Id): tg.Error {
	return typeof data === "string"
		? tg.Error.withId(data)
		: tg.Error.fromData(data);
}

function isTerminalError(error: unknown): boolean {
	return error instanceof tg.Error || error instanceof ProtocolError;
}

async function retryDelay(attempt: number): Promise<void> {
	let delay = Math.min(0.01 * 2 ** Math.min(attempt, 7), 1);
	await tg.sleep(delay);
}

class Channel<T> implements AsyncIterableIterator<T> {
	#closed = false;
	#values: Array<T> = [];
	#waiters: Array<(result: IteratorResult<T>) => void> = [];

	close(): void {
		if (this.#closed) {
			return;
		}
		this.#closed = true;
		while (this.#waiters.length > 0) {
			this.#waiters.shift()!({ done: true, value: undefined });
		}
	}

	next(): Promise<IteratorResult<T>> {
		let value = this.#values.shift();
		if (value !== undefined) {
			return Promise.resolve({ done: false, value });
		}
		if (this.#closed) {
			return Promise.resolve({ done: true, value: undefined });
		}
		return new Promise((resolve) => this.#waiters.push(resolve));
	}

	push(value: T): boolean {
		if (this.#closed) {
			return false;
		}
		let waiter = this.#waiters.shift();
		if (waiter === undefined) {
			this.#values.push(value);
		} else {
			waiter({ done: false, value });
		}
		return true;
	}

	return(): Promise<IteratorResult<T>> {
		this.close();
		return Promise.resolve({ done: true, value: undefined });
	}

	[Symbol.asyncIterator](): AsyncIterableIterator<T> {
		return this;
	}
}
